src/Services/Azure/Authentication/CustomChainedCredential.cs

```diff
@@ -119,4 +140,66 @@ private static DefaultAzureCredential CreateDefaultCredential(string? tenantId)
 
         return new DefaultAzureCredential(defaultCredentialOptions);
     }
+
+    private static TokenCredential? CreateVsCodeBrokerCredential(string? tenantId, Microsoft.Extensions.Logging.ILogger<CustomChainedCredential>? logger = null)
+    {
+        const string vsCodeClientId = "aebc6443-996d-45c2-90f0-388ff96faa56";
+        string userProfile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
+        if (string.IsNullOrEmpty(userProfile))
+        {
+            logger?.LogDebug("VS Code Broker Credential -> User profile directory is null or empty. Cannot locate VS Code authRecord.json.");
+            return null;
+        }
+        string authRecordPath = Path.Combine(userProfile, ".azure", "ms-azuretools.vscode-azureresourcegroups", "authRecord.json");
+        if (!File.Exists(authRecordPath))
+        {
+            // Try .Azure if .azure is not present
+            authRecordPath = Path.Combine(userProfile, ".Azure", "ms-azuretools.vscode-azureresourcegroups", "authRecord.json");
+            if (!File.Exists(authRecordPath))
+            {
+                logger?.LogDebug("VS Code Broker Credential -> authRecord.json not found in either .azure or .Azure directory.");
+                return null;
+            }
+        }
+
+        AuthenticationRecord? authRecord;
+        try
+        {
+            using var stream = File.OpenRead(authRecordPath);
+            authRecord = AuthenticationRecord.Deserialize(stream);
+        }
+        catch (Exception ex)
+        {
+            logger?.LogDebug(ex, "VS Code Broker Credential -> Failed to deserialize VS Code authRecord.json");
+            return null;
+        }
+
+        if (authRecord is null)
+        {
+            logger?.LogDebug("VS Code Broker Credential -> Deserialized VS Code AuthenticationRecord is null.");
+            return null;
+        }
+
+        // Validate client ID
+        if (!string.Equals(authRecord.ClientId, vsCodeClientId, StringComparison.OrdinalIgnoreCase))
+        {
+            logger?.LogDebug("VS Code Broker Credential -> VS Code AuthenticationRecord clientId mismatch. Expected {Expected}", vsCodeClientId);
+            return null;
+        }
+
+        // Prefer explicit tenantId, else use from auth record
+        string effectiveTenantId = !string.IsNullOrEmpty(tenantId)
+            ? tenantId
+            : authRecord.TenantId;
+
+        var options = new InteractiveBrowserCredentialBrokerOptions(0)
+        {
+            ClientId = vsCodeClientId,
+            TenantId = effectiveTenantId,
+            AuthenticationRecord = authRecord
+        };
+
+        logger?.LogDebug("VS Code Broker Credential -> Successfully created VS Code Broker Credential");
+        return new InteractiveBrowserCredential(options);
```

Comment:
Out of curiosity, will this credential create an access token with MFA? It might be important for the upcoming change to [ARM](https://learn.microsoft.com/en-us/entra/identity/authentication/concept-mandatory-multifactor-authentication?tabs=dotnet). I'm still trying to fully wrap my head around the implications for our products.

Reply:
> As I understand things
> 
> 1. A user logs in to VS Code using their username/password or WAM
-> Tyler told me VSCode would only use WAM if it is running on windows.
> 2. They receive a non-MFA token
> 3. Each time the agent attempts a create, update, or delete operation on Azure resources, a step-up challenge (e.g., re-sign in / MFA) will be required, causing the script/agent to be interrupted and the existing cached token to be invalidated
-> It is required only for the first challenge. The MFA token will be cached, eliminating the need for further re-auth.

some comments inline. (assuming in the scope of Azure MCP server and az-cli-extension is not used)

Reply:
> Out of curiosity, will this credential create an access token with MFA? It might be important for the upcoming change to [ARM](https://learn.microsoft.com/en-us/entra/identity/authentication/concept-mandatory-multifactor-authentication?tabs=dotnet). I'm still trying to fully wrap my head around the implications for our products.

This request will obtain a token using the default settings.

If MFA is enforced at the tenant level (e.g., MSFT), an MFA token will be retrieved; otherwise, a non-MFA token is obtained.

I am collaborating with the Entra team to explore options that would allow users to proactively opt in to receive an MFA token.

Reply:
Yup, only on Windows until we ship NodeJS WAM BrokeredAuth packages for macOS and Linux. This scenario here is how we got MSAL to fund the Linux package for later this year (Minh-Anh's feature)

Reply:
> If MFA is enforced at the tenant level (e.g., MSFT), an MFA token will be retrieved; otherwise, a non-MFA token is obtained.

OK, this is good to know. @xiangyan99 , continuing on our conversation from Office Hours this week, I wonder if there's a way to configure desktop client sign-ins for first party apps to require MFA sign-ins. This would mean we never need to worry about the challenge loop. I can only find Conditional Access policies as an option for third party apps, and I'm looking through the FPA docs to see if there's anything special there. What do you think?